File: sort/QuickSort/QuickSort.go
package main

func qs(arr []int, lo int, hi int) {
	if lo >= hi {
		return
	}

	pivotIdx := partition(arr, lo, hi)
	qs(arr, lo, pivotIdx-1)
	qs(arr, pivotIdx+1, hi)
}

func partition(arr []int, lo int, hi int) int {
	pivot := arr[hi]

	idx := lo - 1

	for i := lo; i < hi; i++ {
		if arr[i] <= pivot {
			idx++
			tmp := arr[i]
			arr[i] = arr[idx]
			arr[idx] = tmp
		}
	}

	idx++
	arr[hi] = arr[idx]
	arr[idx] = pivot

	return idx

}

func quick_sort(arr []int) {
	qs(arr, 0, len(arr)-1)
}

File: sort/QuickSort/QuickSort.py
def qs(arr, lo, hi):
    if (lo >= hi):
        return
    
    pivotIdx = partition(arr, lo, hi)
    qs(arr, lo, pivotIdx - 1)
    qs(arr, pivotIdx +1, hi)
    
def partition(arr, lo, hi):
    pivot = arr[hi]
    
    idx = lo - 1
    
    for i in range(lo, hi):
        if (arr[i] <= pivot):
            idx += 1
            tmp = arr[i]
            arr[i] = arr[idx]
            arr[idx] = tmp
        
def quick_sort(arr):
    qs(arr, 0, len(arr) -1)